File: LP1/Assembler/pass1/pass1.py
import re as regex
import json
# Directives mapping
directives = {
    "start" : "(AD, 01)",
    "end" : "(AD, 02)",
    "origin" : "(AD, 03)",
    "equ" : "(AD, 04)",
    "ltorg" : "(AD, 05)"
}  

# Mnemonics size mapping
mneumonicSize = {
    "stop" : 1,
    "add" : 1,
    "sub" : 1,
    "mult" : 1,
    "mover" : 1,
    "movem" : 1,
    "comp" : 1,
    "bc" : 1,
    "div" : 1,
    "read" : 1,
    "print" : 1,
    "start" : 1,
    "end" : 1,
    "origin" : 1,
    "equ" : 1,
    "ltorg" : 1,
    "dc" : 1,
    "ds" : 1
}

# Mnemonics opcode mapping
mneumonic = {
    "stop" : "(IS, 00)",
    "add" : "(IS, 01)",
    "sub" : "(IS, 02)",
    "mult" : "(IS, 03)",
    "mover" : "(IS, 04)",
    "movem" : "(IS, 05)",
    "comp" : "(IS, 06)",
    "bc" : "(IS, 07)",
    "div" : "(IS, 08)",
    "read" : "(IS, 09)",
    "print" : "(IS, 10)",
    "dc" : "(DL, 01)",
    "ds" : "(DL, 02)"
}

# Registers mapping
registers = {
    "areg" : 1,
    "breg" : 2,
    "creg" : 3,
    "dreg" : 4
}

# Conditional codes mapping
conditionalCodes = {
    "lt" : 1,
    "le" : 2,
    "eq" : 3,
    "gt" : 4,
    "ge" : 5,
    "any" : 6
}

# Regular expression pattern for splitting lines
pattern = r'\s+'

# Initialize variables
label = ""
instruction = ""
op1 = ""
op2 = ""
op1code = ""
op2code = ""
current = 0
previous = 0
flag = False        
relativeAddresses = []
IC = []
stCnt = 1
ltCnt = 1

# Symbol and literal tables
symbolTable = {}
literalTable = {}

# Open input and output files
file = open('LP1/Assembler/pass1/input/program.asm','r')
outputfile = open('LP1/Assembler/pass1/output/IntermediateCode.txt','a')

# Iterate through each line in the input file
for line in file:
    label = ""
    instruction = ""
    op1 = ""
    op2 = ""
    op1code = ""
    op2code = ""
    var = ""
    
    # Skip empty lines
    if line == '/n':
        continue
    line = line.strip()

    # Split the line into tokens using the regex pattern
    cmd = regex.split(pattern, line.rstrip())
    cmd = list(map(lambda x: x.lower(), cmd))  # Convert tokens to lowercase

    # Determine the number of tokens and parse accordingly
    if len(cmd) == 4:
        label = cmd[0]
        instruction = cmd[1]
        op1 = cmd[2]
        op2 = cmd[3]
    
    elif len(cmd) == 3:
        cmdIndex = -1
        for c in cmd:
            if c in mneumonic:
                if c == "dc" or c == "ds":
                    var = cmd[0]
                    cmdIndex = 1
                    op1 = cmd[2]
                    flag = True
                    break
                else:
                    cmdIndex = cmd.index(c)
                    break

            if c in directives:
                cmdIndex = cmd.index(c)
                break

        instruction = cmd[cmdIndex]
        if cmdIndex == 0:
            op1 = cmd[0]
            op2 = cmd[1]
        elif flag == False:
            op1 = cmd[cmdIndex-1]
            op2 = cmd[cmdIndex+1]

    elif len(cmd) == 2:
        cmdIndex = -1
        for c in cmd:
            if c in directives:
                cmdIndex = cmd.index(c)
            if c in mneumonic:
                cmdIndex = cmd.index(c)
            instruction = cmd[cmdIndex]
            if cmdIndex == 0:
                op1 = cmd[1]
            else:
                label = cmd[0]

    else:
        instruction = cmd[0]

    # Handle directives
    if instruction in directives:
        if instruction == 'start':
            current = int(cmd[1])
            opcode = directives.get(instruction)
            op1code = f"(C,{op1})"
            outputfile.write(f" {opcode} {op1code}\n")
            continue
        elif instruction == 'end':
            op1code = directives.get(instruction)
            outputfile.write(f"   {opcode}\n")
            break
        elif instruction == 'origin':
            op1code = directives.get(instruction)
            op1 = cmd[1]
            if '-' in op1:
                label = op1.split('-')[0]
                offset = op1.split('-')[1]
                op1code = f"(S, {symbolTable.get(label)[0]})-{offset}"
                previous = current
                current = symbolTable.get(label)[2] - int(offset)
                relativeAddresses.append(previous)
                outputfile.write(f"    {opcode} {op1code}\n")
            elif '+' in op1:
                label = op1.split('+')[0]
                offset = op1.split('+')[1]
                op1code = f"(S, {symbolTable.get(label)[0]})+{offset}"
                previous = current
                current = symbolTable.get(label)[2] + int(offset)
                relativeAddresses.append(previous)
                outputfile.write(f"    {opcode} {op1code}\n")
            else:
                op1code = f"(S, {symbolTable.get(op1)[0]})"
                previous = current
                current = symbolTable.get(op1)[2]
                relativeAddresses.append(previous)
                outputfile.write(f"    {opcode} {op1code}\n")

        # elif instruction == 'equ':
        #     op1 = cmd[0]
        #     op2 = cmd[2]
        #     symbolTable[op1][2] = symbolTable[op2][2]

        elif instruction == 'ltorg':
            for literal, [index, lt, value] in literalTable.items():
                if value == -1:
                    previous = current
                    current += 1
                    relativeAddresses.append(previous)
                    literalTable[literal][2] = previous 
                    opcode = "(DL, 01)"
                    op1code = f"(C, {lt})"  
                    outputfile.write(f"{previous} {opcode} {op1code}\n") 
                else:
                    pass

    # Handle mnemonics
    elif instruction in mneumonic:
        if instruction == 'ds':
            op2code = f""
        if instruction == 'dc':
            op2code = f""
        opcode = mneumonic.get(instruction)
        size = int(mneumonicSize.get(instruction))

        previous = current
        current += size

        relativeAddresses.append(previous)

        # Update symbol table
        if label:
            if label in symbolTable:
                symbolTable[label][2] = previous
            else:
                symbolTable[label] = [stCnt, label, previous]
                stCnt += 1  

        # Process operands
        if instruction == 'bc':
            op1code = f'({conditionalCodes.get(op1)})'

        elif op1.isdigit():
            op1code = f'(C,{op1})'
        elif op1 in registers:
            op1code = f'({registers.get(op1)})'
        elif "=" in op1:
            literal = op1.split('=')[1][1]
            if op1 in literalTable:
                op1code = f"(L, {literalTable.get(op1)[0]})"
            else:
                literalTable[ltCnt] = [ltCnt, literal, -1]
                op1code = f"(L, {ltCnt})"
                ltCnt += 1
        elif instruction == 'ds' or instruction == 'dc':
            symbolTable[op1][2] = previous
        else:
            if op1 in symbolTable:
                op1code = f"(S, {symbolTable.get(op1)[0]})"
            elif op1:
                symbolTable[op1] = [stCnt, op1, -1]
                op1code = f"(S, {stCnt})"
                stCnt += 1
        
        # Process second operand
        if op2.isdigit():
            op2code = f'(C, {op2})'

        elif instruction == 'ds' or instruction == 'dc':
            # print("hello")
            pass
        elif op2 in registers:
            op2code =  f"({registers.get(op2)})"
            
        elif "=" in op2:
            literal = op2.split('=')[1][1]
            if op2 in literalTable:
                op2code = f"(L, {literalTable.get(op2)[0]})"    
            else:
                literalTable[ltCnt] = [ltCnt, literal, -1]
                op2code = f"(L, {ltCnt})"
                ltCnt += 1

        else:
            if op2 in symbolTable:
                op2code = f"(S, {symbolTable.get(op2)[0]})"
            elif op2 and instruction != 'dc':
                symbolTable[op2] = [stCnt, op2, previous]
                op2code = f"(S, {stCnt})"
                stCnt += 1
        
        # Append the instruction with operands to the instruction code list
        if instruction != 'stop':
            IC.append((opcode, op1code, op2code))
            outputfile.write(f"{previous} {opcode} {op1code} {op2code}\n")
        else:
            IC.append((opcode))
            outputfile.write(f"{previous} {opcode}\n")

    else:
        # If instruction is neither a directive nor a mnemonic, exit with an error
        print(instruction, "Instruction not defined. Exiting the program...")
        exit(0)

# Process literals that haven't been assigned a value yet
for literal, [index, lt, value] in literalTable.items():
    if value == -1:
        previous = current
        current += 1
        relativeAddresses.append(previous)
        literalTable[literal][2] = previous 

        opcode = "(DL, 01)"
        op1code = f"(C, {lt})"   
        outputfile.write(f"{previous} {opcode} {op1code}\n")

# Output symbol and literal tables
# print("SymbolTable is: ", symbolTable)
# print("LiteralTable is: ", literalTable)
outputfile.write(f"SymbolTable: {json.dumps(symbolTable)}")
outputfile.write(f"LiteralTable:{json.dumps(literalTable)}")
# Close the input and output files
file.close()
outputfile.close()

File: LP1/Assembler_cpp/Pass1/Pass1.cpp
#include <iostream>
#include <map>
#include <fstream>
#include <vector>
#include <iomanip>
using namespace std;

class SymTab
{
private:
    int id;
    string label;
    int add;

public:
    SymTab() : id(0), label(""), add(0) {}
    SymTab(int id, string label) : id(id), label(label), add(0) {}
    friend class Assembler;
};

class LTab
{
private:
    int id;
    string label;
    int add;

public:
    LTab() : id(0), label(""), add(0) {}
    LTab(int id, string label) : id(id), label(label), add(0) {}
    friend class Assembler;
};

class Assembler
{
private:
    map<string, pair<string, string>> mot;
    map<string, int> len;
    map<string, string> reg;

public:
    Assembler()
    {
        mot = {{"STOP", {"IS", "00"}}, {"ADD", {"IS", "01"}}, {"SUB", {"IS", "02"}}, {"MULT", {"IS", "03"}}, {"MOVER", {"IS", "04"}}, {"MOVEM", {"IS", "05"}}, {"COMP", {"IS", "06"}}, {"BC", {"IS", "07"}}, {"DIV", {"IS", "08"}}, {"READ", {"IS", "09"}}, {"PRINT", {"IS", "10"}}, {"START", {"AD", "01"}}, {"END", {"AD", "02"}}, {"ORIGIN", {"AD", "03"}}, {"EQU", {"AD", "04"}}, {"LTORG", {"AD", "05"}}, {"DC", {"DL", "01"}}, {"DS", {"DL", "02"}}};

        len = {{"IS", 1}, {"AD", 0}, {"DL", 0}};
        reg = {{"AREG", "01"}, {"BREG", "02"}, {"CREG", "03"}, {"DREG", "04"}};
    }

    vector<string> tokenize(string s)
    {
        vector<string> v;
        string x;
        for (auto i : s)
        {
            if (i != ' ')
                x += i;
            else
            {
                v.push_back(x);
                x.clear();
            }
        }
        if (!x.empty())
            v.push_back(x);
        return v;
    }

    SymTab &findsym(vector<SymTab> &s, const string &label)
    {
        static SymTab notFound(-1, "NULL");
        for (auto &i : s)
            if (i.label == label)
                return i;
        return notFound;
    }

    void pass1()
    {
        ifstream f("./input/input.txt");
        ofstream ic("./output/output.txt"), sym("./output/SymTab.txt"), lt("./output/LitTab.txt");
        vector<SymTab> symbolTable;
        vector<LTab> literalTable;

        if (!f.is_open())
        {
            cerr << "Error opening file" << endl;
            return;
        }

        int s_id = 0, l_id = 0, lc = 0;
        string s;

        while (getline(f, s))
        {
            string label, instruction, operand1, operand2, op1code, op2code, write;
            vector<string> tokens = tokenize(s);

            if (tokens.size() == 4)
            {
                label = tokens[0], instruction = tokens[1], operand1 = tokens[2], operand2 = tokens[3];
                SymTab st(s_id++, label);
                st.add = lc;
                symbolTable.push_back(st);
                write += to_string(lc) + " " + "(S," + to_string(st.id) + ") ";
                op1code = mot[instruction].first, op2code = mot[instruction].second;
                write += "(" + op1code + "," + op2code + ") ";
                if (reg.find(operand1) != reg.end())
                {
                    write += "(RG," + reg[operand1] + ") ";
                    if (operand2[0] != '=')
                    {
                        SymTab st1(s_id++, operand2);
                        symbolTable.push_back(st1);
                        write += "(S," + to_string(st1.id) + " ) ";
                    }
                    else
                    {
                        LTab lt1(l_id++, operand2);
                        literalTable.push_back(lt1);
                        write += "(L," + to_string(lt1.id) + " ) ";
                    }
                }
                lc++;
            }
            else if (tokens.size() == 3)
            {
                if (mot.find(tokens[0]) != mot.end())
                {
                    instruction = tokens[0], operand1 = tokens[1], operand2 = tokens[2];
                    op1code = mot[instruction].first, op2code = mot[instruction].second;
                    write += to_string(lc) + " " + "(" + op1code + "," + op2code + ") ";
                    if (reg.find(operand1) != reg.end())
                    {
                        write += "(RG," + reg[operand1] + ") ";
                        if (operand2[0] != '=')
                        {
                            SymTab st1(s_id++, operand2);
                            symbolTable.push_back(st1);
                            write += "(S," + to_string(st1.id) + " ) ";
                        }
                        else
                        {
                            LTab lt1(l_id++, operand2);
                            literalTable.push_back(lt1);
                            write += "(L," + to_string(lt1.id) + " ) ";
                        }
                    }
                    lc++;
                }
                else
                {
                    operand1 = tokens[0], instruction = tokens[1], operand2 = tokens[2];
                    SymTab &st = findsym(symbolTable, operand1);
                    if (st.id == -1)
                    {
                        SymTab s2(s_id++, operand1);
                        s2.add = lc;
                        symbolTable.push_back(s2);
                        st = s2;
                    }
                    st.add = lc;
                    write += to_string(lc) + " (S," + to_string(st.id) + ") ";
                    op1code = mot[instruction].first, op2code = mot[instruction].second;
                    write += "(" + op1code + "," + op2code + ") (C," + operand2 + ")";
                    lc++;
                }
            }
            else if (tokens.size() == 2)
            {
                instruction = tokens[0], label = tokens[1];
                if (instruction == "START")
                {
                    lc = stoi(label);
                    op1code = mot[instruction].first, op2code = mot[instruction].second;
                    write = "\t(" + op1code + "," + op2code + ") (C," + label + ")";
                }
                else if (instruction == "ORIGIN")
                {
                    op1code = mot[instruction].first, op2code = mot[instruction].second;
                    write = to_string(lc) + " (" + op1code + "," + op2code + ") ";
                    int offset = stoi(label.substr(label.find_first_of("+-") + 1));
                    string symbol = label.substr(0, label.find_first_of("+-"));
                    SymTab &st = findsym(symbolTable, symbol);
                    lc = (label.find('+') != string::npos) ? st.add + offset : st.add - offset;
                    write += "(C," + to_string(lc) + ")";
                }
                else
                {
                    op1code = mot[instruction].first, op2code = mot[instruction].second;
                    write = to_string(lc) + " (" + op1code + "," + op2code + ")";
                    SymTab &s1 = findsym(symbolTable, label);
                    write += "(S," + to_string(s1.id) + ")";
                    lc++;
                }
            }
            else
            {
                instruction = tokens[0];
                if (instruction == "LTORG")
                {
                    for (auto &i : literalTable)
                    {
                        write += to_string(lc) + " (AD,05) (DL,01) ";
                        write += "(C," + i.label.substr(2, i.label.size() - 3) + ")\n";
                        i.add = lc++;
                    }
                }
                else
                {
                    op1code = mot[instruction].first, op2code = mot[instruction].second;
                    write = to_string(lc) + " (" + op1code + "," + op2code + ")";
                    lc++;
                }
            }
            ic << write << endl;
        }

        for (auto &i : symbolTable)
            sym << i.id << " " << i.label << " " << i.add << endl;
        for (auto &i : literalTable)
            lt << i.id << " " << i.label << " " << i.add << endl;
    }
};

int main()
{
    Assembler a;
    a.pass1();
    return 0;
}